### PR TITLE
Notifications server

### PR DESCRIPTION
**Pull request by jakehobbs**

### CODE REVIEW

**Comment by jakehobbs** [comment]
I believe this is ready to merge, but I am waiting for reviews.

File: expo/notifications.go
Comment (AlexisDeschamps) [inline expo/notifications.go]: What kind of modifications were made?

File: notifications-worker.go
Comment (AlexisDeschamps) [inline notifications-worker.go]: Is this an error state? Would it be clearer to commit nothing?

File: notifications-worker.go
Comment (AlexisDeschamps) [inline notifications-worker.go]: So this doesn't need to be part of the transaction?

File: notifications-worker.go
Comment (AlexisDeschamps) [inline notifications-worker.go]: Don't we need to provide the transaction level? Or is the default adequate?

File: model/notification.go
Comment (AlexisDeschamps) [inline model/notification.go]: Are you planning on reusing these functions? It might be easier to coordinate the rollbacks if the code is inlined.

**Review by AlexisDeschamps — APPROVED** [review]
LGTM if the SQL queries are working as expected. Did you try making the workers run concurrently and looking for rolled-back transactions or something like that?

File: expo/notifications.go
Comment (jakehobbs) [reply expo/notifications.go]: The only substantial changes are saving the receipt ID & using the Expo Push Token to authorize the request. Maybe it would be better to just fork it and make those changes, but I just copy and pasted to save time.

File: notifications-worker.go
Comment (jakehobbs) [reply notifications-worker.go]: Technically it does not, but it will not execute due to transaction lock if it doesn't use it.

File: notifications-worker.go
Comment (jakehobbs) [reply notifications-worker.go]: It is not an error state. It just means that there are notifications that need sending. I'm not sure what you mean by "commit nothing." Do you mean just skip the rollback b/c nothing would have been selected?

File: notifications-worker.go
Comment (jakehobbs) [reply notifications-worker.go]: This function basically just executes a "begin" statement. Then when we do the "select ... for update", it seems to lock what it needs to so that nothing else can select it.

**Comment by jakehobbs** [comment]
> Did you try making the workers run concurrently and looking for rolled-back transactions or something like that?

@AlexisDeschamps I tested the queries in a SQL shell and it seemed to work as expected. I'll test in prod (since the app is running load-balanced on two servers) once this is merged, and roll back if there are any issues. The app is not public yet, so it will be fine to test in prod.

File: model/notification.go
Comment (jakehobbs) [reply model/notification.go]: Probably not. I just thought it was a bit easier to follow w/ it broken up. I don't feel strongly though.

File: expo/notifications.go
Comment (mdempsky) [inline expo/notifications.go]: You can write this as:

```
switch r.Details["error"] {
case ErrorDeviceNotRegistered:
  ...
case ErrorMessageTooBig:
  ...
...
}
```

File: expo/notifications.go
Comment (mdempsky) [inline expo/notifications.go]: nit: Most of the references to `err` are as `*err` instead. I'd probably change this to `err := PushResponseError{...}` and then just change the return statement to `return &err`.

File: expo/notifications.go
Comment (mdempsky) [inline expo/notifications.go]: These strings are all related, so I'd put them in a single `const` block like:

```
const (
  // SuccesStatus is ...
  SuccessStatus = "ok"

  // ErrorDeviceNotRegistered indicates ...
  ErrorDeviceNotRegistered = "DeviceNotRegistered"

  ...
)
```

You might also consider adding a `type PushStatus string` type for type-safety, but only if it doesn't cause more problems than it helps.

File: expo/notifications.go
Comment (mdempsky) [inline expo/notifications.go]: nit: error message strings conventionally start with a lower-case letter

File: expo/notifications.go
Comment (mdempsky) [inline expo/notifications.go]: I'd recommend making callers provide this as an argument. Ideally only package main calls os.Getenv.

File: model/notification.go
Comment (mdempsky) [inline model/notification.go]: Are these two separate queries in a single statement? Do they need a semicolon to separate them?

File: model/notification.go
Comment (mdempsky) [inline model/notification.go]: Does sqlx have a function for running something as a transaction? If not, I'd write something like:

```
func transact(db *sqlx.DB, do func(tx *sqlx.Tx) error) error {
    tx := db.MustBegin()
    if err := do(tx); err != nil {
      tx.Rollback()
      return err
    }
    return tx.Commit()
}
```

then you can rewrite the code below more like:

```
transact(db, func(tx *sqlx.Tx) error {
    if err := ...; err != nil {
        return err
    }
    ...
    return nil
})
```

File: model/notification.go
Comment (mdempsky) [inline model/notification.go]: FYI, I don't think you need to call tx.Rollback() if tx.Commit() fails.

File: notifications-worker.go
Comment (mdempsky) [inline notifications-worker.go]: nit: remove the blank lines between the github.com imports; typically you just have two import blocks: the standard library and non-standard packages

**Review by mdempsky — COMMENTED** [review]
First pass of superficial comments.

File: model/notification.go
Comment (jakehobbs) [reply model/notification.go]: It's a single query. https://dev.mysql.com/doc/refman/8.0/en/insert-select.html

File: model/notification.go
Comment (jakehobbs) [reply model/notification.go]: Ok, thanks. Will remove that.

File: expo/notifications.go
Comment (jakehobbs) [reply expo/notifications.go]: I'm going to just import a branch of this instead.

File: model/notification.go
Comment (mdempsky) [reply model/notification.go]: TIL, thanks.

Comment (jakehobbs) [reply model/notification.go]: NP. Might be more obvious if I indented instead of having a blank line.

File: model/notification.go
Comment (jakehobbs) [reply model/notification.go]: I don't believe it does, but good thinking.

File: notifications-worker.go
Comment (mdempsky) [reply notifications-worker.go]: I think Alexis means that since it's not a failure, you could just tx.Commit() here instead. Since no side-effects have happened, I don't think there should be any actual difference though.

File: notifications-worker.go
Comment (mdempsky) [inline notifications-worker.go]: I'd recommend trying to stick to database operations during any given transaction. Making external API calls like this can lead to long-lived transactions, which I don't think databases tend to like and can increase the risk of spurious transaction commit failures.

Instead, the usual way I've seen queue-y things like this get implemented is in three phases, something like:

1. First SQL transaction finds a batch of messages to send, and atomically changes all of their statuses from "available" to "leased until {{now + 5 minutes}}". (Tip: Make the "available" status actually "leased until 1970", so you don't actually need to handle multiple cases.)
2. Send all the notifications; make sure this doesn't take more 5 minutes (e.g., use context.WithDeadline or whatever, and probably add some extra slack to account for clock skew).
3. Do another SQL transaction to update all of the messages that were successfully handled as "sent"; optionally (but recommended) change all of the failed messages back to "available".

This limits the risk of duplicate messages to some hard failure happening during step 2, whereas the current design would lead to duplicate messages if any of the transaction fails. The current design also wouldn't work with multiple processes trying to send messages at the same time.

File: notifications-worker.go
Comment (jakehobbs) [reply notifications-worker.go]: Ah, that makes sense.

File: expo/notifications.go
Comment (jakehobbs) [reply expo/notifications.go]: I'm using the official expo package now for this functionality.

File: expo/notifications.go
Comment (jakehobbs) [reply expo/notifications.go]: I'm using the official expo package now for this functionality.

File: expo/notifications.go
Comment (jakehobbs) [reply expo/notifications.go]: I'm using the official expo package now for this functionality.

File: model/notification.go
Comment (jakehobbs) [reply model/notification.go]: This is done in the next commit.

File: expo/notifications.go
Comment (jakehobbs) [reply expo/notifications.go]: I'm using the official expo package now for this functionality.

File: expo/notifications.go
Comment (jakehobbs) [reply expo/notifications.go]: I'm using the official expo package now for this functionality.

File: notifications-worker.go
Comment (jakehobbs) [reply notifications-worker.go]: Next commit will not use a tx for this.

File: model/notification.go
Comment (jakehobbs) [reply model/notification.go]: Next commit will not pass around any transactions.

File: notifications-worker.go
Comment (jakehobbs) [reply notifications-worker.go]: I implemented most of this in the latest commit.

File: model/notification.go
Comment (mdempsky) [inline model/notification.go]: I recommend having the caller pass in `now, deadline time.Time` arguments, rather than making the logic itself tied to the current time. That would make it easier to write unit tests, and it'll also make it easier if you want to have other code use `context.WithDeadline` to make sure the deadline times actually match up.

File: model/notification.go
Comment (mdempsky) [inline model/notification.go:66]: Aside: Does this still need to be two separate queries in a transaction? Can it be a single update/select query?

File: model/user.go
Comment (mdempsky) [inline model/user.go]: Make this a transaction?

File: model/user.go
Comment (mdempsky) [inline model/user.go]: Return errors to caller?

File: notifications-worker.go
Comment (mdempsky) [inline notifications-worker.go]: As-is is fine; but instead of `continue`, I'd probably structure this as:

```
switch {
case r.Status == expo.SuccessStatus:
    ...
case r.Details["error"] == expo.ErrorDeviceNotRegistered:
    ...
default:
    ...
}
```

I think this helps emphasize that there's a multiway branch happening here.

You can also do `if/else if`, but I personally tend to prefer `switch` over `else if`.

File: notifications-worker.go
Comment (mdempsky) [inline notifications-worker.go]: This call seems somewhat redundant with the model.UpdateNotificationStatus call below? Is it possible for model.UpdateNotificationStatus to handle this too, so only one round-trip to the DB is needed?

File: notifications-worker.go
Comment (mdempsky) [inline notifications-worker.go]: I would recommend that. Ideally a context with context.WithDeadline, and then also use the same context for committing the database updates below.

File: model/notification.go
Comment (mdempsky) [inline model/notification.go]: I'm not sure I understand why this query is separate. Maybe add a comment explaining why it's here, and why it's part of the transaction?

**Review by mdempsky — APPROVED** [review]

File: model/notification.go
Comment (jakehobbs) [reply model/notification.go]: Understood. Will do this in next commit.

File: model/notification.go
Comment (jakehobbs) [reply model/notification.go:66]: I thought about that for a bit when designing this, but I couldn't come up with a way to do it in one statement in MySQL. MySQL doesn't have a RETURNING clause for UPDATE statements like Postgres does.

File: model/user.go
Comment (jakehobbs) [reply model/user.go]: Ah, yes. I overlooked that.

File: notifications-worker.go
Comment (jakehobbs) [reply notifications-worker.go]: Making this change.

File: notifications-worker.go
Comment (jakehobbs) [reply notifications-worker.go]: Part of it is redundant. I will rework it a bit.

File: model/notification.go
Comment (jakehobbs) [reply model/notification.go]: It's purpose is just to mark the announcement as "sent" in the announcements table, which has two purposes:
- So it can be displayed on the admin UI that the announcement was queued to be sent
- To skip over sent announcements in the first query

I guess it doesn't actually need to be a part of the transaction. It just needs to not run if the first query does not successful complete.

File: notifications-worker.go
Comment (jakehobbs) [reply notifications-worker.go]: Implemented context in a2f1c3e.

**Pull request merged**
